Common/BuildPreparer.cs

```diff
@@ -99,6 +110,27 @@ private string GetCurrentCommitHash(string moduleName)
             }
         }
 
+        public static List<Dep> GetOptimizedForParallelBuildOrder(Dictionary<Dep, List<Dep>> graph)
+        {
+            var optimizedDeps = new HashSet<Dep>();
+            while (true)
+            {
+                var depsCanRunInParallel = graph.Keys
+                    .Where(dep => !optimizedDeps.Contains(dep) && graph[dep].All(d => optimizedDeps.Contains(d)))
+                    .ToList();
+                if (depsCanRunInParallel.Count == 0) break;
+                foreach (var dep in depsCanRunInParallel)
+                {
+                    optimizedDeps.Add(dep);
+                }
+            }
+            if (graph.Keys.Count > optimizedDeps.Count)
+            {
+                throw new CementException("Unable to build! Circular dependency found!");
+            }
+            return optimizedDeps.ToList();

```

Comment:
Seems you should split graph to chunks, like List<List<Dep>>, where one chunk can be run in parallel.

Reply:
Please remove all logic from BuildDeps file.
Prepare chunks here, and than just iterate over them and jush call build all modules in one chunk.

Reply:
You also can remove `GetTopologicallySortedGraph` and build chunks with `max_size=1`.
And remove two-way of building modules in `BuildDeps`.